File: ansi-vt100-lab/lib/ansi.js
'use strict';

/**
 * Minimal ANSI / VT100 escape sequence toolkit.
 *
 * Every terminal escape sequence starts with the ESC character (0x1B, "\x1b").
 * Most of the sequences we care about are "CSI" (Control Sequence Introducer)
 * sequences, which look like: ESC [ <params> <final-byte>
 *
 *   \x1b[   <- CSI
 *   31      <- parameter(s), e.g. "31" = red foreground
 *   m       <- final byte, "m" means "SGR: Select Graphic Rendition"
 *
 * This file has no dependencies on purpose: everything here is a plain
 * string built by hand, so you can see exactly what bytes are being sent
 * to the terminal. Real projects would reach for a library like `chalk`
 * or `ansi-escapes`, but the point of this lab is to understand what
 * those libraries do under the hood.
 */

const ESC = '\x1b';
const CSI = `${ESC}[`;

// ---------------------------------------------------------------------------
// Cursor movement
// ---------------------------------------------------------------------------

const cursor = {
  up: (n = 1) => `${CSI}${n}A`,
  down: (n = 1) => `${CSI}${n}B`,
  forward: (n = 1) => `${CSI}${n}C`,
  back: (n = 1) => `${CSI}${n}D`,

  // Move cursor to the start of the line, n lines down/up.
  nextLine: (n = 1) => `${CSI}${n}E`,
  prevLine: (n = 1) => `${CSI}${n}F`,

  // Move to an absolute column on the current line (1-based).
  toColumn: (n = 1) => `${CSI}${n}G`,

  // Move to an absolute (row, column) position (1-based, top-left is 1,1).
  to: (row = 1, col = 1) => `${CSI}${row};${col}H`,

  save: `${ESC}7`, // DECSC: save cursor position + attributes
  restore: `${ESC}8`, // DECRC: restore what DECSC saved

  hide: `${CSI}?25l`,
  show: `${CSI}?25h`,
};

// ---------------------------------------------------------------------------
// Erasing
// ---------------------------------------------------------------------------

const erase = {
  // Erase in Display (ED): 0 = cursor->end, 1 = start->cursor, 2 = whole screen
  screenDown: `${CSI}0J`,
  screenUp: `${CSI}1J`,
  screen: `${CSI}2J`,

  // Erase in Line (EL): 0 = cursor->end, 1 = start->cursor, 2 = whole line
  lineEnd: `${CSI}0K`,
  lineStart: `${CSI}1K`,
  line: `${CSI}2K`,
};

// ---------------------------------------------------------------------------
// SGR (Select Graphic Rendition) - colors and text styles
// ---------------------------------------------------------------------------

const style = {
  reset: `${CSI}0m`,
  bold: `${CSI}1m`,
  dim: `${CSI}2m`,
  italic: `${CSI}3m`,
  underline: `${CSI}4m`,
  inverse: `${CSI}7m`,

  // Standard 8/16-color foreground/background (30-37, 90-97 / 40-47, 100-107)
  fg: (code) => `${CSI}${30 + code}m`,
  bg: (code) => `${CSI}${40 + code}m`,
  fgBright: (code) => `${CSI}${90 + code}m`,
  bgBright: (code) => `${CSI}${100 + code}m`,

  // 256-color palette: ESC[38;5;<n>m (fg) / ESC[48;5;<n>m (bg), n = 0-255
  fg256: (n) => `${CSI}38;5;${n}m`,
  bg256: (n) => `${CSI}48;5;${n}m`,

  // 24-bit "true color": ESC[38;2;<r>;<g>;<b>m
  fgRgb: (r, g, b) => `${CSI}38;2;${r};${g};${b}m`,
  bgRgb: (r, g, b) => `${CSI}48;2;${r};${g};${b}m`,
};

// ---------------------------------------------------------------------------
// Alternate screen buffer (what vim/less/htop use so they don't clobber
// your shell scrollback)
// ---------------------------------------------------------------------------

const altScreen = {
  enter: `${CSI}?1049h`,
  exit: `${CSI}?1049l`,
};

module.exports = { ESC, CSI, cursor, erase, style, altScreen };

File: ansi-vt100-lab/internal/ansi/ansi.go
// Package ansi is a minimal, dependency-free ANSI / VT100 escape sequence
// toolkit built for learning purposes.
//
// Every terminal escape sequence starts with the ESC byte (0x1B). Most of
// the sequences used here are "CSI" (Control Sequence Introducer)
// sequences, which look like:
//
//	ESC [ <params> <final-byte>
//	\x1b[   31       m
//	 ^      ^        ^
//	 CSI    param    "m" = SGR (Select Graphic Rendition)
//
// Every function in this package just concatenates plain strings, so you
// can see exactly what bytes get sent to the terminal. Real projects would
// reach for something like github.com/gdamore/tcell or
// github.com/charmbracelet/lipgloss, but the point of this lab is to
// understand what those libraries do under the hood.
package ansi

import "fmt"

const (
	Esc = "\x1b"
	CSI = Esc + "["
)

// --- Cursor movement --------------------------------------------------

// CursorUp/Down/Forward/Back move the cursor n cells in that direction.
func CursorUp(n int) string      { return fmt.Sprintf("%s%dA", CSI, n) }
func CursorDown(n int) string    { return fmt.Sprintf("%s%dB", CSI, n) }
func CursorForward(n int) string { return fmt.Sprintf("%s%dC", CSI, n) }
func CursorBack(n int) string    { return fmt.Sprintf("%s%dD", CSI, n) }

// CursorNextLine/PrevLine move to the start of the line, n lines down/up.
func CursorNextLine(n int) string { return fmt.Sprintf("%s%dE", CSI, n) }
func CursorPrevLine(n int) string { return fmt.Sprintf("%s%dF", CSI, n) }

// CursorToColumn moves to an absolute column on the current line (1-based).
func CursorToColumn(n int) string { return fmt.Sprintf("%s%dG", CSI, n) }

// CursorTo moves to an absolute (row, col) position (1-based; top-left is 1,1).
func CursorTo(row, col int) string { return fmt.Sprintf("%s%d;%dH", CSI, row, col) }

const (
	CursorSave    = Esc + "7" // DECSC: save cursor position + attributes
	CursorRestore = Esc + "8" // DECRC: restore what DECSC saved

	CursorHide = CSI + "?25l"
	CursorShow = CSI + "?25h"
)

// --- Erasing ------------------------------------------------------------

// Erase in Display (ED).
const (
	EraseScreenDown = CSI + "0J" // cursor -> end of screen
	EraseScreenUp   = CSI + "1J" // start of screen -> cursor
	EraseScreen     = CSI + "2J" // whole screen
)

// Erase in Line (EL).
const (
	EraseLineEnd   = CSI + "0K" // cursor -> end of line
	EraseLineStart = CSI + "1K" // start of line -> cursor
	EraseLine      = CSI + "2K" // whole line
)

// --- SGR: colors and text styles ----------------------------------------

const (
	Reset     = CSI + "0m"
	Bold      = CSI + "1m"
	Dim       = CSI + "2m"
	Italic    = CSI + "3m"
	Underline = CSI + "4m"
	Inverse   = CSI + "7m"
)

// FG/BG select the standard 8-color foreground/background (code 0-7:
// black, red, green, yellow, blue, magenta, cyan, white).
func FG(code int) string { return fmt.Sprintf("%s%dm", CSI, 30+code) }
func BG(code int) string { return fmt.Sprintf("%s%dm", CSI, 40+code) }

// FGBright/BGBright select the bright variants of the same 8 colors.
func FGBright(code int) string { return fmt.Sprintf("%s%dm", CSI, 90+code) }
func BGBright(code int) string { return fmt.Sprintf("%s%dm", CSI, 100+code) }

// FG256/BG256 select from the 256-color palette (n = 0-255).
func FG256(n int) string { return fmt.Sprintf("%s38;5;%dm", CSI, n) }
func BG256(n int) string { return fmt.Sprintf("%s48;5;%dm", CSI, n) }

// FGRGB/BGRGB select a 24-bit "true color".
func FGRGB(r, g, b int) string { return fmt.Sprintf("%s38;2;%d;%d;%dm", CSI, r, g, b) }
func BGRGB(r, g, b int) string { return fmt.Sprintf("%s48;2;%d;%d;%dm", CSI, r, g, b) }

// --- Alternate screen buffer ---------------------------------------------
//
// What vim/less/htop use so they don't clobber your shell scrollback.

const (
	AltScreenEnter = CSI + "?1049h"
	AltScreenExit  = CSI + "?1049l"
)
